lua_modules/ds18b20/ds18b20.lua

```diff
@@ -41,7 +44,49 @@ function setup(dq)
   if(pin == nil) then
     pin = defaultPin
   end
-  ow.setup(pin)
+  ow.setup(pin) 
+end
+temp_result = 0
+
+function callBack(addr, func)
+  -- reset last temp
+  temp_result = 0
+  print(addr:byte(1,9))
+  if checkAddrCrc(addr) then
+    if ((addr:byte(1) == 0x10) or (addr:byte(1) == 0x28)) then
+      -- print("Device is a DS18S20 family device.")
+      parasite = parasitePowered(addr)
+      ow.reset(pin)
+      ow.select(pin, addr)
+      ow.write(pin, 0x44, 1)
+      if parasite then
+        -- Parasite powered, need to wait ~750ms
+        print ("Using parasite power wait")
+        -- TODO: use different wait than delay
+        tmr.alarm(1, 750, 0, function () processRead(addr, func) end)
+      else
+        -- external powered, waiting for finish
+        print ("Using external powered wait")
+        repeat
+          local res = ow.read(pin)
+          tmr.wdclr()
+        until ( res ~= 0 )
+        processRead(addr, func)
+      end
+    end
+  end
+end
+
+function processRead(addr, callback_func)
+  data = readScratchpad(addr)
+  if (checkDataCrc(data)) then
+    temp_result = convertTemperature(data, addr, unit)
+  else
+    temp_result = "Data CRC is not valid"
+  end
+  tmr.wdclr()
```

Comment:
Oh,
I thought that Pull request will stay only on that 1st commit, now I just
using it as backup of work while other work :) I not used github recently,
only local gitlab, so I not knew it will add comments there aswell.
Yes you are right, I just trying rewrite it somehow from procedural to
event one, but it have to be refatored really all also there is things
like, as long it will be async with call backs, it allows you to call that
callback even previous is not processed yet, so I have to do some queue or
make some "false" return while calling too fast or so...
I will think about that later when I will not be at work.

Petr

2016-07-15 15:51 GMT+02:00 Marcel Stör <notifications@github.com>:

> In lua_modules/ds18b20/ds18b20.lua
> <https://github.com/nodemcu/nodemcu-firmware/pull/1400#discussion_r70975808>
> :
>
> > +          tmr.wdclr()
> > +        until ( res ~= 0 )
> > +        processRead(addr, func)
> > +      end
> > +    end
> > +  end
> > +end
> > +
> > +function processRead(addr, callback_func)
> > +  data = readScratchpad(addr)
> > +  if (checkDataCrc(data)) then
> > +    temp_result = convertTemperature(data, addr, unit)
> > +  else
> > +    temp_result = "Data CRC is not valid"
> > +  end
> > +  tmr.wdclr()
>
> http://nodemcu.readthedocs.io/en/latest/en/modules/tmr/#tmrwdclr:
>
> In general, if you ever need to use this function, you are doing it wrong.
>
> 😉
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodemcu/nodemcu-firmware/pull/1400/files/71a82a017da556c78b507c28443bfbdbcaeb99ca#r70975808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGIWvzdAEMbR8TqYRmRQQvfs-56pWxg_ks5qV5B1gaJpZM4JM1no>
> .
>



-- 
Petr Kracik クラチク ペトル
Mobil: +420 602 707 704
Skype: petr.kracik
E-Mail: petrkr@petrkr.net
Web http://www.petrkr.net